Feat: React 19 Upgrade

kierancap opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Please make sure you're familiar with and follow the instructions in the contributing guidelines (found in the CONTRIBUTING.md file).

Please fill out the information below to expedite the review and (hopefully) merge of your pull request!
-->

### Why

Closing https://github.com/pmndrs/react-spring/issues/2341

### What

Main changes:

- Removed all usage of `forwardRef` - As per the React docs, will be fully deprecated in future.
- Replaced all `ref` fields with the valid React Types - Any polymorphic components were set as `any` for the `RefObject` generic.
- Added valid initial values for all usages of `useRef` which were missing inits.
- Refactored the `SpringContext` to resemble a more sensible Context structure

### Checklist

- [x] Removed forwardRef
- [x] Added correct `ref` fields
- [x] Added initial values for `useRef`
- [x] Refactored SpringContext

<!-- if you untick ready to be merged & you haven't submitted as a draft, we will change it to draft. -->

<!-- feel free to add additional comments -->

### Additional Comments

This was based on this PR: https://github.com/pmndrs/react-spring/pull/2349

As the PR and issue have been open for a good while now, it's obvious that no further development was going to happen - I'm currently blocked by this library, so I thought it'd try and help with moving this forward.


### PR Diff

I'm unsure as to why the diff is heavily focused on formatting changes - I was sure to run prettier via the NPM script before submitting this PR. Happy to outline all the direct code changes if necessary.

kierancap committed:
feat: Bumped deps and refactored usages of forwardRef

kierancap committed:
feat: disabled workflows

kierancap committed:
feat: Refactored SpringContext

kierancap committed:
feat: Reverted testing change

kierancap committed:
formatting: useSprings

kierancap committed:
Merge pull request #2 from Capventis/1-feat-react-19

feat: Bumped deps and refactored usages of forwardRef

kierancap committed:
ci: enabled workflows again

thomazcapra commented:
👀

abriginets commented:
@joshuaellis apologies for reaching out directly, but if you have some time, could you please take a look at this PR? We'd really appreciate your review when it’s convenient for you. Thanks a lot!

johansunden commented:
Is @joshuaellis the only person who can review this? Would be great to get this in and not have it end up like [the other attempt at upgrading to React 19](https://github.com/pmndrs/react-spring/pull/2349)

kierancap commented:
> Is @joshuaellis the only person who can review this? Would be great to get this in and not have it end up like [the other attempt at upgrading to React 19](https://github.com/pmndrs/react-spring/pull/2349)

Yeah... I was hoping that the fact this PR passed all the tests would mean it'd get reviewed sooner rather than later. At this rate someone may as well fork this library fully and re-distribute it. Shambolic from the maintainers if I'm honest

joshuaellis commented:
Hey, firstly – sorry I recently started a family & I also have a FT job so unfortunately, open-source software is taking a back seat. Secondly, thanks for the PR! Here's some follow up questions:

* Have you tried running this library with a react16|17|18 app? I'm wondering if these changes to forwardRef will mean this is a breaking change to the library?
* I think your changes to SpringContext are probably breaking changes because previously `<SpringContext>` was possible – now it's not? Maybe i'm misunderstanding the change though.
* "Shambolic from the maintainers if I'm honest" – how many maintainers do you think react-spring has?

Also, some other things i think we're missing from this is:
* updating the package.json of the packages to add react19 as an acceptable peer-dep
* There's a console warning about act in the useTransition test, be good to tackle that

Otherwise, this looks super promising!

kierancap commented:
Appreciate the reply, great to hear from you. Congratulations on the family news!

I'd be happy to test out those changes regarding older version of React once I get the opportunity.

The context change is breaking due to the nature of changing the Provider to be separate from the instantiated context itself. It follows the Context Provider pattern; [Here's](https://kentcdodds.com/blog/how-to-use-react-context-effectively) an article around it - I find it to be way better for DX. If you'd prefer it to remain as it was previously, it may involve a bit more in-depth tinkering due to the type errors generated from the current approach. I'm not torn on either approach - As long as the resulting code is readable and maintainable for future development. I can take a look at refactoring the types and structure to potentially allow for a single object to safely contain all the necessary functionality.

In terms of my comment about the maintenance of this project, it's more targeted to the complete lack of communication towards the community. The main issue attached to this PR has hundreds of people waiting for this change, after you had mentioned about updating it in December. Granted I appreciate your new family changes meant you were not available, but I do think that the lack of communication severely impacted this whole process.

As I was forced to make this change due to a work dependancy on a 1st party library, I cannot guarantee that I'll have time to look at this in the short term. I'll do my best.

Hopefully I can have time next week to delve deeper into the good points you mentioned.

I'll try to keep this PR up to date with that

joshuaellis commented:
> The context change is breaking due to the nature of changing the Provider to be separate from the instantiated context itself. It follows the Context Provider pattern; [Here's](https://kentcdodds.com/blog/how-to-use-react-context-effectively) an article around it - I find it to be way better for DX. If you'd prefer it to remain as it was previously, it may involve a bit more in-depth tinkering due to the type errors generated from the current approach. I'm not torn on either approach - As long as the resulting code is readable and maintainable for future development. I can take a look at refactoring the types and structure to potentially allow for a single object to safely contain all the necessary functionality.

tbf i'm kind of happy to remove it, I don't think _anyone_ uses it. Also it breaks everything in RCS for imo not a lot of gain...

> Hopefully I can have time next week to delve deeper into the good points you mentioned.

sounds good! ✨

kierancap commented:
Just had a read through some other known React 19 OSS projects - They seem to be using `forwardRef` still - Even though the React team has specifically mentioned it being removed in future and is currently labeled as deprecated. In that case I'll revert my forwardRef changes as it should still function for the current React 19 version - Will get to that next week 😃

dagatsoin commented on docs/app/components/Header/HeaderSidePanel.tsx:
```diff
@@ -10,7 +10,6 @@ import { MenuDocs } from '../Menu/MenuDocs'
 
 import { NavigationSchema } from '../../../scripts/docs/navigation'
 import { SiteThemePicker } from '../Site/SiteThemePicker'
-import { forwardRef } from 'react'
```
breaking change: this may break react version < 19.

According to the react [doc](https://react.dev/blog/2024/12/05/react-19#ref-as-a-prop), forwardRef is still supported until further notice. We should keep it.

kierancap replied:
Yep agreed - I plan to revert these forwardRef changes when I can!

kierancap replied:
This is now resolved - I've re-added the forwardRefs :)

dagatsoin commented on package.json:
```diff
@@ -95,8 +95,8 @@
     "pretty-quick": "4.0.0",
-    "react": "18.3.1",
-    "react-dom": "18.3.1",
+    "react": "19.0.0",
```
deps: is locking to 19.0.0 not to restrictive? What about accepting at least patch or even minor version offset? (same question for the type)

kierancap replied:
I was mainly following the previous setup - As it was locked to that specific React 18 version, happy to change it though

dagatsoin replied:
Ho my bad, I did not see it was a dev dependencies.
However, all the packages' peerDependencies are needed to be updated to React 19 (I assume ^19.0.0 ?)

kierancap replied:
Yep - I've now added React 19^ as a peer dep to all the packages :)

dagatsoin reviewed (changes requested):
Hi, just a quick first pass. I stopped on the forwardRef as it has already a big impact. Tell me your though.

kierancap committed:
revert: reverted forwardRef removal & updated peer deps

kierancap committed:
fix: removed act batching and used act for timer advancing and flushing tasks

kierancap commented:
Right people - I've now made these changes:

- [x] Reverted removal of forwardRef (yuck)
- [x] Added React 19 as a Peer Dep to all the packages
- [x] Fixed the `act()` error from the tests

DominikScholz commented:
Would be awesome if this could be merged in!, only thing missing for my company to move to react 19, any way I can help?

smollweide commented:
@kierancap 

**First of all, thank you for the pull request!**
However, it seems that the `ref` is not working as expected.

**Issue**
The `ref` should point to the DOM element so that we can access its properties like getBoundingClientRect(). Currently, it's null.

**Example**
Here's a minimal example to illustrate the issue. The button's width should be displayed when it's clicked:

```tsx
export function RefAnimated() {
  const ref = useRef<HTMLButtonElement>(null);
  const [_, setCounter] = useState(0);
  const Button = animated('button');
  
  return (
    <Button
      ref={ref}
      onClick={() => setCounter((v) => v + 1)}
    >
      With ref (width:{' '}
      {ref.current?.getBoundingClientRect().width || 0}px)
    </Button>
  );
}
```
_When clicking the button, the width should update, but currently it does not behave as expected._

nathan-knight commented on packages/animated/src/withAnimated.tsx:
```diff
@@ -27,7 +27,7 @@ export const withAnimated = (Component: any, host: HostConfig) => {
     (Component.prototype && Component.prototype.isReactComponent)
 
-  return forwardRef((givenProps: any, givenRef: Ref<any>) => {
+  return (givenProps: any, givenRef: Ref<any>) => {
```
Shouldn't this still use `forwardRef`?

kierancap replied:
Good spot! Resolved in this commit: https://github.com/pmndrs/react-spring/pull/2363/commits/8d61c6da8b6ed8f286d8c95b1d052fba4b009b22

kierancap committed:
fix: readded forwardRef in withAnimated

kierancap commented:
> @kierancap
> 
> **First of all, thank you for the pull request!** However, it seems that the `ref` is not working as expected.
> 
> **Issue** The `ref` should point to the DOM element so that we can access its properties like getBoundingClientRect(). Currently, it's null.
> 
> **Example** Here's a minimal example to illustrate the issue. The button's width should be displayed when it's clicked:
> 
> ```tsx
> export function RefAnimated() {
>   const ref = useRef<HTMLButtonElement>(null);
>   const [_, setCounter] = useState(0);
>   const Button = animated('button');
>   
>   return (
>     <Button
>       ref={ref}
>       onClick={() => setCounter((v) => v + 1)}
>     >
>       With ref (width:{' '}
>       {ref.current?.getBoundingClientRect().width || 0}px)
>     </Button>
>   );
> }
> ```
> 
> _When clicking the button, the width should update, but currently it does not behave as expected._

Hey! Thank you for taking the time to test this - As mentioned in a comment left after you, I had missed a forwardRef on the `withAnimated` function. Which may have caused this issue. Would you mind retesting this with the latest commit? https://github.com/pmndrs/react-spring/pull/2363/commits/8d61c6da8b6ed8f286d8c95b1d052fba4b009b22

Thank you!

kierancap commented:
Hey @joshuaellis - Apologies for the direct ping!

Just wondering of next steps for this? Are you happy to override the reviewing/approval or do I need to wait for re-review from the previous reviewers? 

Thanks mate!

joshuaellis commented:
I'm OOO till Tuesday, but if the CI passes we can cut a beta, people can test it and report back any issues? Give it a couple of weeks before going to stable.

kierancap commented:
> I'm OOO till Tuesday, but if the CI passes we can cut a beta, people can test it and report back any issues? Give it a couple of weeks before going to stable.

Sounds like a plan! The CI is currently failing early due to me not pushing an updated lock (post updating the peer deps). I'll update that today - Then on tuesday we can see about doing that beta release! 

Thank you

EDIT: Have pushed that lock file fix - Hopefully will let the CI pass properly (Can't tell until we try it!)

VincentHui commented:
glorious PR! I've watched and hoped for a long time and just want to say thanks. I'm checking out this branch and giving it a look just in case

kierancap committed:
fix: updated lock file

ivtpz commented:
Thanks for making this update!

Is there a beta version now to test with?

bustinjurkz commented:
this would be awesome to merge !

joshuaellis commented:
Might be worth following this pr: #2368 from what i can tell, both contributors want to continue there? 🤷🏼‍♀️

dagatsoin commented:
@abriginets  @johansunden  @DominikScholz  @smollweide  @VincentHui  @ivtpz
thx for your patience all. The v10.0.0-beta.0 is published.

And yes @johansunden, this PR won't be update, all @kierancap work was put in #2368  ([cf this discussion](https://github.com/pmndrs/react-spring/pull/2368#issuecomment-2822159713))
